confdgnmi/src/confd_gnmi_api_adapter.py

```diff
@@ -59,32 +63,183 @@ def get_inst(cls) -> GnmiServerAdapter:
+                return _confd.ITER_RECURSE
+
+            cdb.diff_iterate(sub_sock, sub_point, iter, 0, None)
+            self.put_event(self.SubscriptionEvent.SEND_CHANGES)
+            log.debug("self.change_db=%s", self.change_db)
+            log.debug("<==")
+
+        def process_changes(self):
+            log.debug("==>")
+            # make subscription for all self.monitored_paths
+            sub_sock = socket(AF_INET, SOCK_STREAM, 0)
+            prio = 10
+
+            cdb.connect(sub_sock, cdb.SUBSCRIPTION_SOCKET, '127.0.0.1',
+                        _confd.CONFD_PORT)
+            for p in self.monitored_paths:
+                log.debug("subscribing config p=%s", p)
+                # TODO hash - breaks generic usage
+                # TODO for now we subscribe path for both, config and oper,
+                # TODO subscribe only for paths that exist
+                # it may be more efficient to find out type of path and subscribe
+                # only for one type
+                cdb.subscribe2(sub_sock, cdb.SUB_RUNNING, 0, prio,
+                               0, p)
+                log.debug("subscribing operational p=%s", p)
+                cdb.subscribe2(sub_sock, cdb.SUB_OPERATIONAL, 0, prio,
+                               0, p)
+            cdb.subscribe_done(sub_sock)
+            log.debug("subscribe_done")
+            assert self.stop_pipe is not None
+            _r = [sub_sock, self.stop_pipe[0]]
+            _w = []
+            _e = []
+            try:
+                while True:
+                    log.debug("_r=%s", _r)
+                    r, w, e = select.select(_r, _w, _e)
+                    log.debug("r=%s", r)
+                    if self.stop_pipe[0] in r:
+                        v = os.read(self.stop_pipe[0], 1)
+                        assert v == b'x'
+                        log.debug("Stopping ConfD loop")
+                        break
+                    if sub_sock in r:
+                        try:
+                            list = cdb.read_subscription_socket2(sub_sock)
+                            for s in list[2]:
+                                self.process_subscription(sub_sock, s)
+                            cdb.sync_subscription_socket(sub_sock,
+                                                         cdb.DONE_PRIORITY)
+                        except _confd.error.Error as e:
+                            # Callback error
+                            if e.confd_errno is _confd.ERR_EXTERNAL:
+                                log.exception(e)
+                            else:
+                                raise e
+
+            except Exception as e:
+                log.exception(e)
+
+            log.debug("<==")
+
+        def start_monitoring(self):
+            log.debug("==>")
+            assert self.change_thread is None
+            assert self.stop_pipe is None
+            log.debug("** creating change_thread")
+            self.stop_pipe = os.pipe()
```

Comment:
1. Is the pipe ever closed? Does not look so.
2. I understand why you need this, but isn't there a way to get along without the pipe? Maybe there is not, just it is a bit strange to send random characters just to tell the thread to stop.

Reply:
I did not find any other way, maybe it exists. It is not stopping thread directly, but triggering `select` in (ConfD) loop.
I'll add close.

Reply:
I agree, it does not look like there is a better way. It just hurts a bit to read - or write - code so cumbersomely (*is that a word?*) mixing sockets and events.

Reply:
sockets and file descriptors

Reply:
1. added closing of pipe (line 229 in `stop_monitoring`)
            os.close(self.stop_pipe[0])
            os.close(self.stop_pipe[1])